projects/dqn/src/algorithms.py

```diff
@@ -0,0 +1,175 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2022, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy as np
+import torch
+
+from .misc import Scheduler
+from .networks import DQNNetwork
+
+
+class AlgorithmBase():
+    pass
+
+
+class DQN(AlgorithmBase):
+    """
+    DQN algorithm implementation compatible with Vernon
+    Parts of the code adapted from:
+    https://github.com/transedward/pytorch-dqn/blob/master/dqn_learn.py
+    """
+
+    def __init__(
+        self,
+        observation_size,
+        num_actions,
+        num_steps,
+        device,
+        q_function_network_class=DQNNetwork,
+        exploration_rate_scheduler_class=Scheduler,
+        q_function_optimizer=torch.optim.Adam,
+        discount_rate=0.99,
+        q_function_lr=1e-2,
+        exploration_rate=5e-2,
+        exploration_rate_final=1e-3,
+        target_function_update_rate=1e-2,
+        target_function_update_frequency=100,
+        clip_error=True,
+    ):
+
+        self.device = device
+
+        # Initialize networks
+        input_channels = observation_size[2]
+        self.q_function = q_function_network_class(
+            input_channels=input_channels,
+            num_actions=num_actions
+        )
+        self.q_function.to(self.device)
+        self.target_q_function = q_function_network_class(
```

Comment:
Supposed to be initialized from exact same weights.

Reply:
changed to deepcopy